Prototype of legacy configuration split

Pierstoval opened the pull request:
TODO:

- [x] Move most of the includes to DI-based config
- [ ] Split them as much as possible
- [x] Make the config be set up by the Kernel when handling an HTTP request
- [ ] Make the console Application use the Kernel
- [ ] Make the config be set up by the Kernel when handling a Console command
- [ ] Make sure the config is set up by the Kernel when running an Install (should be automatic, but must be checked)
- [ ] Make sure the config is set up by the Kernel when running an Update (should be automatic, but must be checked)

Pierstoval committed:
Prototype of legacy configuration split

cedric-anne commented on src/Glpi/Config/LegacyConfigurators/BaseConfig.php:
```diff
@@ -0,0 +1,274 @@
+
+        include_once(GLPI_ROOT . "/inc/autoload.function.php");
+
+        // Check if web root is configured correctly
+        if (!isCommandLine()) {
+            $included_files = array_filter(
+                get_included_files(),
+                function (string $included_file) {
+                    // prevent `tests/router.php` to be considered as initial script
+                    return realpath($included_file) !== realpath(sprintf('%s/tests/router.php', GLPI_ROOT));
+                }
+            );
+
+            $initial_script = array_shift($included_files) ?? __FILE__;
+
+            // If `auto_prepend_file` configuration is used, ignore first included files
+            // as long as they are not located inside GLPI directory tree.
+            $prepended_file = ini_get('auto_prepend_file');
+            if ($prepended_file !== '' && $prepended_file !== 'none') {
+                $prepended_file = stream_resolve_include_path($prepended_file);
+                while (
+                    $initial_script !== null
+                    && !str_starts_with(
+                        realpath($initial_script) ?: '',
+                        realpath(GLPI_ROOT)
+                    )
+                ) {
+                    $initial_script = array_shift($included_files);
+                }
+            }
+
+            if (realpath($initial_script) !== realpath(sprintf('%s/public/index.php', GLPI_ROOT))) {
+                echo sprintf(
+                    'Web server root directory configuration is incorrect, it should be "%s". See installation documentation for more details.' . PHP_EOL,
+                    realpath(sprintf('%s/public', GLPI_ROOT))
+                );
+                exit(1);
+            }
+        }
```
This can be removed now. As long as all the legacy bootstraping logic is removed from `inc/xxx.php` files, it is not anymore possible to execute GLPI without using the `public/index.php` file.
```suggestion
```

cedric-anne commented on src/Glpi/Config/LegacyConfigurators/BaseConfig.php:
```diff
@@ -0,0 +1,274 @@
+
+    public function execute(Request $request): void
+    {
+        if (!defined('GLPI_ROOT')) {
+            define('GLPI_ROOT', $this->projectDir);
+        }
```
It should not be defined anywhere else. If it does, it must fail so we can detect other unexpected declarations.
```suggestion
        define('GLPI_ROOT', $this->projectDir);
```

Pierstoval committed:
Update src/Glpi/Config/LegacyConfigurators/BaseConfig.php

Removing code that's now useless since we only use the public-based front controller

Co-authored-by: Cédric Anne <cedric.anne@gmail.com>

Pierstoval committed:
Update src/Glpi/Config/LegacyConfigurators/BaseConfig.php

Co-authored-by: Cédric Anne <cedric.anne@gmail.com>

Pierstoval commented:
Closing in favor of https://github.com/glpi-project/glpi/pull/17278
